src/tmate.py

```diff
@@ -56,25 +63,61 @@ class FingerprintError(Exception):
     """Represents an error with generating fingerprints from public keys."""
 
 
-def install_dependencies() -> None:
-    """Install dependenciese required to start tmate-ssh-server container.
+def _setup_docker(proxy_config: typing.Optional[state.ProxyConfig] = None) -> None:
+    """Install and configure proxy settings for docker if available.
+
+    Args:
+        proxy_config: The proxy configuration to enable for dockerd.
 
     Raises:
-        DependencySetupError: if there was something wrong installing the apt package
-            dependencies.
+        DependencySetupError: if there was a problem installing/setting up Docker.
     """
+    if proxy_config:
+        environment = jinja2.Environment(
+            loader=jinja2.FileSystemLoader("templates"), autoescape=True
+        )
+        docker_template = environment.get_template("docker_daemon.json.j2")
+        daemon_config = docker_template.render(
+            HTTP_PROXY=proxy_config.http_proxy,
+            HTTPS_PROXY=proxy_config.https_proxy,
+            NO_PROXY=proxy_config.no_proxy,
+        )
+        DOCKER_DAEMON_CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
+        DOCKER_DAEMON_CONFIG_PATH.touch(exist_ok=True)
+        DOCKER_DAEMON_CONFIG_PATH.write_text(daemon_config, encoding="utf-8")
+
     try:
-        apt.update()
-        apt.add_package(APT_DEPENDENCIES)
-    except (apt.PackageNotFoundError, apt.PackageError) as exc:
-        raise DependencySetupError("Failed to install apt packages.") from exc
+        apt.add_package("docker.io", update_cache=True)
+    except (apt.PackageNotFoundError, apt.PackageError, subprocess.CalledProcessError) as exc:
+        raise DependencySetupError("Failed to apt install Docker.") from exc
     passwd.add_group("docker")
     try:
         passwd.add_user_to_group(USER, "docker")
     except ValueError as exc:
         raise DependencySetupError(f"Failed to add user {USER} to docker group.") from exc
 
 
+def install_dependencies(proxy_config: typing.Optional[state.ProxyConfig] = None) -> None:
+    """Install dependenciese required to start tmate-ssh-server container.
+
+    Args:
+        proxy_config: The proxy configuration to enable for dockerd.
+
+    Raises:
+        DependencySetupError: if there was something wrong installing the apt package
+            dependencies.
+    """
+    try:
+        apt.add_package(APT_DEPENDENCIES, update_cache=True)
+    except (apt.PackageNotFoundError, apt.PackageError, subprocess.CalledProcessError) as exc:
+        raise DependencySetupError("Failed to install apt packages.") from exc
+    try:
+        _setup_docker(proxy_config=proxy_config)
```

Comment:
Why not putting `_setup_docker` in the same try bock as `apt.add_package(APT_DEPENDENCIES, update_cache=True)` and handle all the exceptions together? If _setup_docker reraises the original exceptions in line 91 , the except block in line 112 can be shared for both

Reply:
It's been suggested before that try excepts should be in the smallest scope possible.

Reply:
Where and why? I don't think there's a one fits all solution for these situations. Why capturing the exceptions in two different lines when you can enhance mainteinability using a single except block?

Reply:
Changed!